Switch dataframe constructor to use dispatch

saulshanabrook opened the pull request:
This is an attempt to add extensibility to the `DataFrame` constructor so that third party libraries can register their own ways of converting to a Pandas Dataframe. It does this by creating a [`singledispatch` function](https://docs.python.org/3/library/functools.html#functools.singledispatch) that is used in the constructor.

For example, Dask could implement the function like this:

```python
from pandas.core.construction import create_dataframe
import dask.datafame

@create_dataframe.register
def _create_dataframe_dask(data: dask.datafame.DataFrame, *args, **kwargs):
    return create_dataframe(data.compute(), *args, **kwargs)
```

Then, if a downstream library tries to construct a Pandas dataframe from a dask dataframe, it will work:

```python
import dask
import pandas

df = dask.datasets.timeseries()

assert isinstance(pandas.DataFrame(df), pandas.DataFrame)
```


This is response to [the thread about providing a protocol for dataframes](https://discuss.ossdata.org/t/a-dataframe-protocol-for-the-pydata-ecosystem/267) to present an alternative for the underlying use case. The alternative is:

1. Force libraries like sk learn to depend on pandas
2. Have them called `pandas.Dataframe` on their input data to see if it can be turned into a dataframe
3. Have third party libraries with alternative dataframe implementations register themselves with this function provided here.

It doesn't try to solve any sort of out-of-core dataframe API conversation and it does require all libraries to have Pandas as a hard dependency.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] Review name of function

saulshanabrook committed:
Switch dataframe constructor to use dispatch

saulshanabrook committed:
blacken

saulshanabrook committed:
flake8

saulshanabrook committed:
mypy fixes

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -36,6 +37,7 @@
 
 import numpy as np
 import numpy.ma as ma
+import numpy.ma.mrecords as mrecords
```
i think we import this lazily to speed up import time

saulshanabrook replied:
I thought maybe that was the case. However, looking at the PR that added the import (https://github.com/pandas-dev/pandas/pull/5579/files) I didn't see any discussion about import time. Do you know if there are any benchmarks to back this up I could run?

If it's a deal breaker, I can move the mrecords check into the ma check like it was in the original code.

datapythonista commented:
I like the idea, the code looks much cleaner and readable (besides being extensible).

Not sure if the pytest errors are caused by your changes (the linting problems are from this PR, the docs is unrelated and should be fixed now).

saulshanabrook committed:
style fixes

saulshanabrook committed:
Merge pandas/master into rework-constructor

jreback commented on pandas/core/frame.py:
```diff
@@ -8548,6 +8462,140 @@ def isin(self, values) -> "DataFrame":
 ops.add_special_arithmetic_methods(DataFrame)
 
 
+@functools.singledispatch
```
if we are moving this i think moving this into
pandas/frame/construction.py would be fine

saulshanabrook replied:
Sounds good.

jreback commented on pandas/core/frame.py:
```diff
@@ -8548,6 +8462,140 @@ def isin(self, values) -> "DataFrame":
+    return create_block_manager({}, *args, **kwargs)
+
+
+@create_block_manager.register
```
not in love with this name

maybe create_dataframe ?

saulshanabrook replied:
Sure I can rename it to that. Might just be a bit confusing though since the return type is a block manager, not a dataframe.

saulshanabrook committed:
rename and take class instead of instance

saulshanabrook committed:
Move create_dataframe to construction

saulshanabrook committed:
Sort imports

saulshanabrook committed:
Fix calling

saulshanabrook committed:
Add test for custom constructor

saulshanabrook committed:
Added whats new

saulshanabrook committed:
unused imports

saulshanabrook commented:
I have moved this to the `construction.py` file, added a test for it, and changed it to just accept the dataframe class, instead of the instance.

The tests are passing locally on my mac with `./test_fast.sh` but they seem to be failing in CI with some error that is maybe hiding the underlying error?

```
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/main.py", line 247, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/xdist/dsession.py", line 112, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/xdist/dsession.py", line 135, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/xdist/dsession.py", line 272, in worker_collectreport
INTERNALERROR>     self._failed_worker_collectreport(node, rep)
INTERNALERROR>   File "/home/vsts/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/xdist/dsession.py", line 302, in _failed_worker_collectreport
INTERNALERROR>     if rep.longrepr not in self._failed_collection_errors:
INTERNALERROR> TypeError: unhashable type: 'ExceptionChainRepr'
```

saulshanabrook committed:
Merge pandas/master into rework-constructor

jreback commented:
can u add a test where you import dask and register this like in your example 

see test_downstream.py for now we skip if it’s not available

TomAugspurger commented:
I'm not opposed to the idea in principle, but I don't think we should be publicly exposing the block manager in any way to downstream libraries, even if it's just for libraries and not users.

I think a top-level `pd.dataframe` function which returns a DataFrame is much friendlier to single dispatc. Then we could have

```python
@singledispatch.register
def dataframe(data: dask.dataframe.DataFrame, index, columns, ...):
    return pd.DataFrame(data.compute(), index, columns, ...)
```

without dask having to know or care what a BlockManager is.

jbrockmendel commented:
> but I don't think we should be publicly exposing the block manager in any way to downstream libraries, even if it's just for libraries and not users.

Seconded.

saulshanabrook commented:
> I'm not opposed to the idea in principle, but I don't think we should be publicly exposing the block manager in any way to downstream libraries, even if it's just for libraries and not users.
> 
> I think a top-level pd.dataframe function which returns a DataFrame is much friendlier to single dispatch.

So not change the existing constructor at all, just add a separate single dispatch function that by default just calls the constructor?

TomAugspurger commented:
Yep.

saulshanabrook commented:
Would that maybe be more confusing for users then, since there are now two public methods of constructing dataframes instead of one?

Also, as you see from the example, the dask authors don't have to know what a block_manager is, because then can just call `create_dataframe` with a dataframe itself.

datapythonista commented:
Would it make sense to do both separately?

What we've got now for the block manager constructor, but private, so the code is cleaner.

And then a public `create_dataframe`, `pandas.dataframe`, the `__init__` itself, or whatever for other libraries to override the `DataFrame` constructor?

TomAugspurger commented:
> Would it make sense to do both separately?

Sure, but separately is the key :) Then we can decide whether we like the singledispatch-style code structure better that the status quo, without having to worry about API discussions. We'll also need to measure performance.

> the __init__ itself, or whatever for other libraries to override the DataFrame constructor?

It's not clear to me how singledispatch works on instance methods like `__init__`, since the first argument is an uninitialized DataFrame.

---

I'd also like a bit more research on who would benefit from a dispatchable `pandas.dataframe` method for creating dataframes. Who would be the consumers of that API? Dask is one candidate, but they've gotten along OK without `pd.DataFrame(dask_dataframe)` working (I recall one issue where a user was confused by it not working).

saulshanabrook commented:
> It's not clear to me how singledispatch works on instance methods like __init__, since the first argument is an uninitialized DataFrame.

There is a [`singledispatchmethod`](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod), that could work if it is just used internally. However, then all ovverrides for it would have to be other methods on the `DataFrame` class so would have to live there not in `constructors.py` and wouldn't be able to be added to by third party libraries (which I guess maybe you want?)

> I'd also like a bit more research on who would benefit from a dispatchable pandas.dataframe method for creating dataframes. Who would be the consumers of that API? Dask is one candidate, but they've gotten along OK without pd.DataFrame(dask_dataframe) working (I recall one issue where a user was confused by it not working).

The supposed user would be downstream libraries like sklearn, so they could call `pd.DataFrame` on their input data and have it be coerced to a pandas dataframe, if possible.

TomAugspurger commented:
Is scikit-learn discussing turning arbitrary objects into pandas dataframes? Currently https://github.com/scikit-learn/enhancement_proposals/pull/37 is discussing pandas in -> pandas out. There's some discussing there around other objects (e.g. xarray DataArray and Dataset) but it seems just as likely for scikit-learn to develop an interface for preserving the type of the input, rather than coercing to a DataFrame.

saulshanabrook commented:
>  it seems just as likely for scikit-learn to develop an interface for preserving the type of the input, rather than coercing to a DataFrame.

Yep that's a possibility. I wanted to open this to show another possibility, in case that was interesting to them to avoid the dataframe interface conversation.

jreback commented:
theoretically a nice idea, but

closing as stale if you want to continue, please open a new PR.

TomAugspurger commented:
Opened https://github.com/pandas-dev/pandas/issues/34799 for discussing the idea of a singly-dispatched `pd.dataframe`.
